## Objective in this notebook is to read the needed .csv's into pandas dfs

In [19]:
import pandas as pd

df1 = pd.read_csv("properties_2017_view.csv")
df2 = pd.read_csv("predictions_2017_view.csv")


In [20]:
# check to see that the csv's were read

df1.head()

id  parcelid  airconditioningtypeid  architecturalstyletypeid  \
0   0  10754147                    NaN                       NaN   
1   1  10759547                    NaN                       NaN   
2   2  10843547                    NaN                       NaN   
3   3  10859147                    NaN                       NaN   
4   4  10879947                    NaN                       NaN   

   basementsqft  bathroomcnt  bedroomcnt  buildingclasstypeid  \
0           NaN          0.0           0                  NaN   
1           NaN          0.0           0                  NaN   
2           NaN          0.0           0                  5.0   
3           NaN          0.0           0                  3.0   
4           NaN          0.0           0                  4.0   

   buildingqualitytypeid  calculatedbathnbr  ...  numberofstories  \
0                    NaN                NaN  ...              NaN   
1                    NaN                NaN  ...              NaN   
2                    NaN                NaN  ...              1.0   
3                    6.0                NaN  ...              1.0   
4                    NaN                NaN  ...              1.0   

   fireplaceflag  structuretaxvaluedollarcnt  taxvaluedollarcnt  \
0            NaN                         NaN                9.0   
1            NaN                         NaN            27516.0   
2            NaN                    660680.0          1434941.0   
3            NaN                    580059.0          1174475.0   
4            NaN                    196751.0           440101.0   

   assessmentyear  landtaxvaluedollarcnt  taxamount  taxdelinquencyflag  \
0            2016                    9.0        NaN                 NaN   
1            2015                27516.0        NaN                 NaN   
2            2016               774261.0   20800.37                 NaN   
3            2016               594416.0   14557.57                 NaN   
4            2016               243350.0    5725.17                 NaN   

   taxdelinquencyyear  censustractandblock  
0                 NaN                  NaN  
1                 NaN                  NaN  
2                 NaN                  NaN  
3                 NaN                  NaN  
4                 NaN                  NaN  

[5 rows x 59 columns]

In [21]:
df2.head()

id  parcelid  logerror transactiondate
0   0  14297519  0.025595      2017-01-01
1   1  17052889  0.055619      2017-01-01
2   2  14186244  0.005383      2017-01-01
3   3  12177905 -0.103410      2017-01-01
4   4  10887214  0.006940      2017-01-01

### But df1's got 59 columns!  Way too many.  Use what I need.

In [23]:
clean_df1 = pd.read_csv("properties_2017_view.csv", usecols=["bathroomcnt", "bedroomcnt", "calculatedfinishedsquarefeet", "fips", "taxvaluedollarcnt", "taxamount"])
clean_df1.head()

# So that you don't have to scroll all the way right, I'm choosing: 
# bathroomcnt
# bedroomcnt
# calculatedfinishedsquarefeet
# fips
# taxvaluedollarcnt
# taxamount

bathroomcnt  bedroomcnt  calculatedfinishedsquarefeet  fips  \
0          0.0           0                           NaN  6037   
1          0.0           0                           NaN  6037   
2          0.0           0                       73026.0  6037   
3          0.0           0                        5068.0  6037   
4          0.0           0                        1776.0  6037   

   taxvaluedollarcnt  taxamount  
0                9.0        NaN  
1            27516.0        NaN  
2          1434941.0   20800.37  
3          1174475.0   14557.57  
4           440101.0    5725.17

### Similarly, I don't need logerror from df2:

In [27]:
clean_df2 = pd.read_csv("predictions_2017_view.csv", usecols=["id", "parcelid", "transactiondate"])
clean_df2.head()

id  parcelid transactiondate
0   0  14297519      2017-01-01
1   1  17052889      2017-01-01
2   2  14186244      2017-01-01
3   3  12177905      2017-01-01
4   4  10887214      2017-01-01

### Combining the dataframes

In [32]:
frames = [clean_df1, clean_df2]
df = pd.concat(frames, sort=True)

#### Make sure they combined:

In [45]:
df.head()

bathroomcnt  bedroomcnt  calculatedfinishedsquarefeet    fips  id  \
0          0.0         0.0                           NaN  6037.0 NaN   
1          0.0         0.0                           NaN  6037.0 NaN   
2          0.0         0.0                       73026.0  6037.0 NaN   
3          0.0         0.0                        5068.0  6037.0 NaN   
4          0.0         0.0                        1776.0  6037.0 NaN   

   parcelid  taxamount  taxvaluedollarcnt transactiondate  
0       NaN        NaN                9.0             NaN  
1       NaN        NaN            27516.0             NaN  
2       NaN   20800.37          1434941.0             NaN  
3       NaN   14557.57          1174475.0             NaN  
4       NaN    5725.17           440101.0             NaN

In [46]:
df.shape

(2000, 9)

In [47]:
df.columns

Index(['bathroomcnt', 'bedroomcnt', 'calculatedfinishedsquarefeet', 'fips',
       'id', 'parcelid', 'taxamount', 'taxvaluedollarcnt', 'transactiondate'],
      dtype='object')

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2000 entries, 0 to 999
Data columns (total 9 columns):
bathroomcnt                     1000 non-null float64
bedroomcnt                      1000 non-null float64
calculatedfinishedsquarefeet    931 non-null float64
fips                            1000 non-null float64
id                              1000 non-null float64
parcelid                        1000 non-null float64
taxamount                       968 non-null float64
taxvaluedollarcnt               868 non-null float64
transactiondate                 1000 non-null object
dtypes: float64(8), object(1)
memory usage: 156.2+ KB


In [49]:
df.isna().sum()

bathroomcnt                     1000
bedroomcnt                      1000
calculatedfinishedsquarefeet    1069
fips                            1000
id                              1000
parcelid                        1000
taxamount                       1032
taxvaluedollarcnt               1132
transactiondate                 1000
dtype: int64

In [50]:
df.calculatedfinishedsquarefeet.value_counts(dropna=False)

NaN       1069
1344.0      15
1440.0      15
1248.0       9
1152.0       9
          ... 
1635.0       1
2622.0       1
1529.0       1
2071.0       1
2194.0       1
Name: calculatedfinishedsquarefeet, Length: 672, dtype: int64

### WHAT??  1,069 NaNs?  That's a lot of missing data.  I gotta address this...

In [51]:
df.taxvaluedollarcnt.value_counts(dropna=False)

NaN          1132
9.0             5
550000.0        3
369854.0        2
350260.0        2
             ... 
146259.0        1
382547.0        1
1092245.0       1
287396.0        1
775000.0        1
Name: taxvaluedollarcnt, Length: 846, dtype: int64

### Same thing here: 1,132 rows of missing data.

In [52]:
df.taxamount.value_counts(dropna=False)

NaN        1032
377.09        1
573.52        1
2051.26       1
5117.52       1
           ... 
742.76        1
6384.54       1
237.30        1
91.60         1
503.36        1
Name: taxamount, Length: 969, dtype: int64

### Again, over a 1,000 rows of missing data.  Time to clean up the NaN values...

In [56]:
# Get booleans to see which values are NaNs; True = NaN

df.isnull() 

bathroomcnt  bedroomcnt  calculatedfinishedsquarefeet   fips     id  \
0          False       False                          True  False   True   
1          False       False                          True  False   True   
2          False       False                         False  False   True   
3          False       False                         False  False   True   
4          False       False                         False  False   True   
..           ...         ...                           ...    ...    ...   
995         True        True                          True   True  False   
996         True        True                          True   True  False   
997         True        True                          True   True  False   
998         True        True                          True   True  False   
999         True        True                          True   True  False   

     parcelid  taxamount  taxvaluedollarcnt  transactiondate  
0        True       True              False             True  
1        True       True              False             True  
2        True      False              False             True  
3        True      False              False             True  
4        True      False              False             True  
..        ...        ...                ...              ...  
995     False       True               True            False  
996     False       True               True            False  
997     False       True               True            False  
998     False       True               True            False  
999     False       True               True            False  

[2000 rows x 9 columns]

In [59]:
# Check to see which rows contain NaN; again, True = has NaN in that row

df.any(axis=1)

0      True
1      True
2      True
3      True
4      True
       ... 
995    True
996    True
997    True
998    True
999    True
Length: 2000, dtype: bool

### Ugh.  They're all over the place...

- I don't want to get rid of them 